### PR TITLE
proxy: validate urls before setting as proxies (SC-170)

### PR DESCRIPTION
**Pull request by orndorffgrant**
## Proposed Commit Message

> proxy: validate urls before setting as proxies
>
> Introduces a proxy validation function that checks two things:
>
> 1. if the proxy value is a parseable url with http or https scheme
> 2. if the proxy can be used to reach important hardcoded destinations
> 
> If either check fails, the current operation is aborted and an error
message is printed.

## Test Steps

1. Configure ua to use an invalid url as a proxy. Ensure enabling any service fails.
2. Configure ua to use a valid url that doesn't point to a functioning proxy. Ensure enabling any service fails.
3. Configure ua to use a valid url that points to a functioning proxy. Ensure enabling any service succeeds.

## Desired commit type::

 - [ ] Squash and merge: Using the "Proposed Commit Message"
 - [ ] Create a merge commit and use "Proposed Commit Message"
 - [x] Rebase and merge, no merge commit, rely only on existing commit messages

## Checklist:

 - [x] I have updated or added any unit tests accordingly
 - [x] I have updated or added any integration tests accordingly
 - [ ] I have updated or added any documentation accordingly

### CODE REVIEW

File: uaclient/util.py
Comment (blackboxsw) [inline uaclient/util.py:819]: When encountering invalid I'd prefer the approach of stopping the line with loud UserFacingErrorr/log errors. In cloud-init we are too flexible and try as hard as we can to ignore non-critical errors, which leads to much later recognition that something is misconfigured on the system. The more strict we are, the more time is saves a customer having invalid config recognized.

File: uaclient/util.py
Comment (blackboxsw) [inline uaclient/util.py:814]: Validate proxy in my mind should succeed or raise exceptions if proxy values are provided. If no values are provided, a return None is in order. I'm welcome to persuasion otherwise.

**Review by blackboxsw — CHANGES_REQUESTED** [review]
suggestion for discussion

File: uaclient/util.py
Comment (orndorffgrant) [reply uaclient/util.py:819]: Yep this is something I wanted to bring up/talk about - this was just my first draft. I do like the "fail fast and loud" approach when appropriate.

Thinking about it more, I think it makes sense here to fail loudly, because just like you say, if we continue after this error then we will try to use entitlements without the proxies in working order. The user specifically configured ua to use a proxy, so continuing operations while effectively ignoring the proxy would be unexpected/undesirable. 

We can chat more about this but I'll probably go ahead and change it to raise a UserFacingError

**Review by lucasmoura — COMMENTED** [review]
LGTM

Just waiting for CI to see if the tests are working as expected

**Comment by orndorffgrant** [comment]
@lucasmoura Thanks for the review! We can't actually merge quite yet because I need to add calls to the validate function as part of `ua config refresh` which was just merged earlier today (thanks for that too :))

**Comment by lucasmoura** [comment]
@orndorffgrant fair enough, makes total sense

**Comment by orndorffgrant** [comment]
@blackboxsw @lucasmoura I added the validate calls to the `ua refresh config` function and updated the integration tests, so this should be ready for review. 

Note that we'll want to call this validation as a part of `ua config set` as well, so if we merge #1696 first, then I'll add that in this PR. Or if we merge this PR first, then we'll need to add validate calls to #1696 before merging.

**Review by lucasmoura — APPROVED** [review]

**Review by blackboxsw — APPROVED** [review]
+ 1 Let's land this now as-is since ths is great & works well, and I have some rebase fixes to re-apply to #1696 anyway. I'll sort both after this merges.

**Comment by blackboxsw** [comment]
CI failures were lxd zfs destroy storage issues.

**Pull request merged**
